### PR TITLE
Numpy-style broadcasting for all mathematical functions

### PR DESCRIPTION
**Pull request by gchanan**
all functions mentioned in: implements all functions mentioned in https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes

### CODE REVIEW

**Comment by gchanan** [comment]
Here is numpy-style broadcasting for pointwise math and reduction functions.

This also changes the keepdim default from True to False.

I have a longer writeup in the works categorizing the numpy semantics and showing that these are the only backwards incompatible changes necessary to get numpy-style broadcasting (i.e. what's left from here only adds functionality that would currently give you an error).

I'll link the writeup when that's done, but feel free to review before then (although you may wish to hold off on merging).

**Comment by soumith** [comment]
let's aim to merge this after the nips deadline on Friday

File: torch/lib/TH/THStorage.c [hunk 1]
```diff
@@ -66,39 +66,76 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
   return copy;
 }
 
-TH_API void THLongStorage_calculateExpandGeometry(long *tensorSizes, long *tensorStrides, long tensorDim, THLongStorage *sizes, long **esz, long **est) {
-  ptrdiff_t ndim = THLongStorage_size(sizes);
-  long numUnsqueezed = ndim - tensorDim;
+TH_API int THLongStorage_inferSize2(THLongStorage *output, long *sizesA, long dimsA, long *sizesB, long dimsB, int raiseErrors) {
```
Comment (ezyang) [inline torch/lib/TH/THStorage.c]: Given the sizes and dimensions of two tensors, write out the final size when the two tensors are broadcasted together.

File: torch/lib/TH/THStorage.c [hunk 2]
```diff
@@ -66,39 +66,76 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
...
+  memcpy(THLongStorage_data(output), expandedSizes, sizeof(long)*ndim);
+  THFree(expandedSizes);
+  return 0;
+}
 
-  for (long i = numUnsqueezed - 1; i > -1; --i) {
-    expandedSizes[i] = 1;
-    expandedStrides[i] = expandedSizes[i+1] * expandedStrides[i+1];
-  }
+TH_API int THLongStorage_inferExpandGeometry(long *tensorSizes, long *tensorStrides, long tensorDim, THLongStorage *sizes, long **esz, long **est, int raiseErrors) {
```
Comment (ezyang) [inline torch/lib/TH/THStorage.c]: Given an existing tensor with size tensorSizes, strides tensorStrides and dimension tensorDim, compute the new expanded size (saved to esz) and new necessary stride (saved to est) necessary to broadcast (expand) this tensor to sizes.

File: torch/lib/TH/generic/THTensor.c [hunk 3]
```diff
@@ -285,21 +285,80 @@ void THTensor_(resize5d)(THTensor *self, long size0, long size1, long size2, lon
...
+  THTensor_(setStorageNd)(r, THTensor_(storage)(tensor), THTensor_(storageOffset)(tensor), THLongStorage_size(sizes), expandedSizes, expandedStrides);
+  THFree(expandedSizes);
+  THFree(expandedStrides);
 
-  THTensor *result = THTensor_(new)();
-  THTensor_(setStorageNd)(result, THTensor_(storage)(tensor), THTensor_(storageOffset)(tensor), THLongStorage_size(sizes), expandedSizes, expandedStrides);
+  return 0;
+}
+
+int THTensor_(expand2)(THTensor *ra, THTensor *rb, THTensor *opa, THTensor *opb, int raiseErrors) {
```
Comment (ezyang) [inline torch/lib/TH/generic/THTensor.c]: Given two tensors, expand them so that they have identical dimensions and sizes, following broadcasting semantics.

File: tools/cwrap/plugins/Broadcast.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
+from . import CWrapPlugin
+from string import Template
+
+
+class Broadcast(CWrapPlugin):
```
Comment (ezyang) [inline tools/cwrap/plugins/Broadcast.py]: Adds boilerplate to broadcast arguments to be identical sizes, before running the algorithm proper. Only works for binary operators.

File: test/test_torch.py [hunk 5]
```diff
@@ -939,6 +939,156 @@ def test_arange(self):
...
+            with warnings.catch_warnings(record=True) as w:
+                r0 = tensorfn(t0_fn, t1)
+                verifyFallbackWarnings(w)
+            with warnings.catch_warnings(record=True) as w:
+                r1 = tensorfn(t1_fn, t0)
+                verifyFallbackWarnings(w)
+            self.assertEqual(t0.size(), r0.size())
+            self.assertEqual(t1.size(), r1.size())
+
+            # case 4: not broadcastable and not nEleme equal -- tested by test_fallback
```
Comment (ezyang) [inline test/test_torch.py]: s/nEleme/nElem/

File: torch/lib/TH/THStorage.c [hunk 6]
```diff
@@ -104,6 +104,49 @@ TH_API int THLongStorage_inferSize2(THLongStorage *output, long *sizesA, long di
   return 0;
 }
 
+TH_API int THLongStorage_inferSizeN(THLongStorage *output, int n, long **sizes, long *dims, int raiseErrors) {
```
Comment (ezyang) [inline torch/lib/TH/THStorage.c]: Given n tensors with given sizes and dims, broadcast all of them to a single size, which is recorded in output.

**Review by ezyang — COMMENTED** [review]
I love it. I went through and wrote some documentation for the new functions while reading. I didn't do a very careful "is this logic correct" code review.

**Comment by gchanan** [comment]
@ezyang thank you, I will incorporate these suggestions.

If you are interested in how this works / future plans, you may find these notes interesting: https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes.

**Comment by albanD** [comment]
@gchanan could it be possible to have a global switch (at compile time if not possible at runtime) that would disable the automatic broadcasting? For example that just disable the cwrap plugin?
I (and maybe other people) prefer to use view/expand and not have the broadcast that can hide subtle bugs.

**Comment by gchanan** [comment]
@albanD good question and I understand the motivation for tightly controlling your own code.  What I'm not sure about is how to make the distinction between your code and library code (or if that distinction is even well defined).

To support what you want as written, we'd have to ensure that our libraries work both with and without broadcasting, which will greatly increase the maintenance burden.  It's also not clear to me the distinction between user and library code is really clear anyway: let's say some NN layer is written in such a way that it e.g. multiples the input tensor by a (4,1) tensor and you pass in a (4) tensor.  Was it library or user code that caused broadcasting?

What I have now is a UserWarning if your code does not broadcast, but uses the old (deprecated) 1-d pointwise operations (you previously only needed nElem to match up for many functions).  I could also add an optional warning for the backwards-incompatible case, that is, the sizes don't match, so you would have used the 1-d pointwise operations, but you are now broadcasting.

It sounds like what you want is a warning if broadcasting changes the sizes at all (I'd be hestitant to make it an error, not a warning, given the argument about library code above).  That shouldn't be too difficult to implement, although I'm not sure how useful it will be if our library code broadcasts a lot (you might get a lot of warnings).  What do you think?

**Comment by albanD** [comment]
Ho right I assumed that our library will stay broadcasting free (as it is right now).
I agree that keeping it that way may be a bit more complex for people that want to send new PR and always use broadcasting. But here the change are quite minor as you would just need to slightly adapt new PR and the existing code will not change and work as is.
I guess @soumith and @apaszke will need to decide if the possibility of disabling the broadcasting is worth this maintaining cost ?

**Comment by pavanky** [comment]
What about a per tensor switch instead of a global switch ? This will however increase the complexity in user code.

**Comment by gchanan** [comment]
If you pass in a non-broadcasting tensor to the library, does it switch to a broadcasting tensor?  If it doesn't, you have the same issue of the library having to work with both (and having to define in what cases the flag is copied).

**Comment by gchanan** [comment]
I think this is ready for review and now implements all functions mentioned in https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes (i.e. it no longer includes only pointwise mathematical and comparison functions like the title says).

Re: turning off broadcasting, the only case I implemented this for was for copy, which allows you to pass a "broadcast" parameter (default: True).  There were a number of places in the code where it was clearly intended to copy tensors as 1-d, which wasn't really true for other functions.

This PR now also includes some warnings that you can enable to detect backwards incompatible changes.  In particular:

1) Setting torch.utils.backcompat.broadcast.warning.enabled=True will cause Python warnings in the case where broadcast occurs but previously 1-d view style pointwise ops occured.

2) Setting torch.utils.backcompat.keepdim.warning.enabled=True will cause Python warnings in the case where the default value of keepdim is used for 1-d reductions.

I've manually enabled these and verified that the only places these warnings are triggered are in the tests are directly in tests, i.e. there are no library calls (at least that are covered by tests) where we rely on the old behavior.  This should make turning on these warnings less noisy.

**Comment by gchanan** [comment]
rebased the commits.

File: docs/source/notes/broadcasting.rst [hunk 7]
```diff
@@ -0,0 +1,108 @@
+.. _broadcasting-semantics:
+
+Broadcasting semantics
+==============
+
+Many PyTorch operations support :any:`NumPy Broadcasting Semantics <numpy.doc.broadcasting>`.
+
+In short, if a PyTorch operation supports broadcast, then its Tensor arguments can be
+automatically expanded to be of equal sizes (without making copies of the data).
```
Comment (killeent) [inline docs/source/notes/broadcasting.rst]: Nit: once and for all I would like us to settle on either "size" or "shape"

File: tools/cwrap/plugins/Broadcast.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+from string import Template
+
+# Arguments to the Broadcast Plugin:
+# broadcast: args_to_broadcast_against [inplace] [fallback]
+# [args_to_broadcast_against]: either a single argument (e.g. "arg1") or a comma-seperated
+#                              list of two arguments (e.g. "tensor1,tensor2") indicating
+#                              arguments to broadcast specified argument (usually "self") against
+# [inplace] will generate code for in-place function, which doesn't allow the in-place
+#           argument to be broadcast
+# [fallback] if tensors aren't broadcastable, preserves "element number" pointwise behavior,
```
Comment (killeent) [inline tools/cwrap/plugins/Broadcast.py]: Nit: make this `broadcast_fallback` or something that makes it explicit what behavior we are "falling back" from.

File: tools/cwrap/plugins/Broadcast.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+#            where only number of elements need to match, and tensors are viewed as 1-dimensional.
+# [dims] if the tensors shouldn't be broadcast to specific tensor or tensors, but a combination
```
Comment (killeent) [inline tools/cwrap/plugins/Broadcast.py]: It's a little clear what `dims` is used for, can you expand upon this.

File: torch/csrc/Module.cpp [hunk 10]
```diff
@@ -466,6 +466,54 @@ PyObject *THPModule_addDocStr(PyObject *_unused, PyObject *args)
   Py_RETURN_NONE;
 }
 
+
+
+PyObject *THPModule_inferSize(PyObject *_unused, PyObject *args)
```
Comment (killeent) [inline torch/csrc/Module.cpp]: Why is this added here? What is it used for?

File: torch/csrc/copy_utils.h [hunk 11]
```diff
@@ -50,24 +51,88 @@ inline PyObject * THPCopyMethod(const THPCopyList& v, PyObject *self, PyObject *
...
+  return self;
+}
+
+inline PyObject * THPTensorCopyMethod(const THPCopyList& v, PyObject *self, PyObject *args, PyObject *kwargs)
+{
+  PyObject *src;
+  PyObject *async = Py_False;
+  PyObject *broadcast = Py_True;
+  static char *kwlist[] = {"source", "async", "broadcast", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O|OO:copy_", kwlist, &src, &async, &broadcast)) {
```
Comment (killeent) [inline torch/csrc/copy_utils.h]: Is this not able to verify the types of keyword arguments? I can't remember.

File: torch/csrc/copy_utils.h [hunk 12]
```diff
@@ -50,24 +51,88 @@ inline PyObject * THPCopyMethod(const THPCopyList& v, PyObject *self, PyObject *
...
-  bool async=false)
+  bool async=false,
+  bool broadcast=true)
 {
-  auto wrapper = [copyFunc](PyObject* dst_, PyObject* src_) {
+  auto wrapper = [copyFunc](PyObject* dst_, PyObject* src_, bool broadcast) {
     TensorDst* dst = THPTypeInfo<TensorDst>::cdata(dst_);
     TensorSrc* src = THPTypeInfo<TensorSrc>::cdata(src_);
 
+    TensorSrc *src_save = src;
```
Comment (killeent) [inline torch/csrc/copy_utils.h]: Can you explain what is going on here with broadcast?

File: torch/csrc/cuda/expand_utils.cpp [hunk 13]
```diff
@@ -0,0 +1,188 @@
+#include "torch/csrc/cuda/THCP.h"
```
Comment (killeent) [inline torch/csrc/cuda/expand_utils.cpp]: Why do we need these files (e.g. the same for csrc below)? Why can't we just call `THTensor_(new)`?

File: torch/csrc/expand_utils.h [hunk 14]
```diff
@@ -0,0 +1,155 @@
+#ifndef THP_EXPAND_UTILS_H
```
Comment (killeent) [inline torch/csrc/expand_utils.h]: Can you document this file? It's very unclear what is going on.

File: torch/csrc/utils.cpp [hunk 15]
```diff
@@ -551,3 +551,33 @@ void THPPointer<THPGenerator>::free() {
 }
 
 template class THPPointer<THPGenerator>;
+
+static bool backCompatBroadcastWarn = false;
```
Comment (killeent) [inline torch/csrc/utils.cpp]: Someone better at C++ than I can evaluate whether or not this works as a global variable?

File: torch/lib/TH/THStorage.c [hunk 16]
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
...
+        }
+      }
+    }
+
+    expandedSizes[ i ] = max_dim_size;
+  }
+  THLongStorage_resize(output, ndim);
+  memcpy(THLongStorage_data(output), expandedSizes, sizeof(long)*ndim);
+  THFree(expandedSizes);
+  return 0;
```
Comment (killeent) [inline torch/lib/TH/THStorage.c]: Above in `THSize_isSameSizeAs` we return 1 on success, 0 on failure. Here we return 0 on success, -1 on failure? It would be nice for these to be consistent?

**Review by killeent — COMMENTED** [review]
In the future I think it would be helpful if we could come up with some sort of mechanism for breaking up PRs into smaller chunks, although I'm not blaming you for that @gchanan.

Due to the large scope of this PR, I didn't really look at the logic so much, I guess I'll just have to trust you. I added a few nits and questions where I didn't understand generally why things were done.

File: docs/source/notes/broadcasting.rst (hunk 7 above)
Comment (gchanan) [inline docs/source/notes/broadcasting.rst]: Agreed, and I wasn't consistent in this PR.  In general I prefer "shape" but there is already a size() function and a torch.Size object.  I generally used size/sizes except in cases I thought shape was significantly clearer.

File: torch/csrc/cuda/expand_utils.cpp (hunk 13 above)
Comment (gchanan) [inline torch/csrc/cuda/expand_utils.cpp]: This is tricky, and I'll add a comment, although it's already done in a number of places in the code.  The functions in expand_utils.h/expand_utils-inl.h are defined in terms of the LIBRARY_STATE and related macros (because the copy functions are defined the same way).  So, there are actually two versions of each function (with and without THCState) and we avoid redefining the non-cuda ones by including THCP.h (that in terms defines the macros) and then including those files.

File: torch/csrc/expand_utils.h (hunk 14 above)
Comment (gchanan) [inline torch/csrc/expand_utils.h]: Same as above.

File: torch/csrc/utils.cpp (hunk 15 above)
Comment (gchanan) [inline torch/csrc/utils.cpp]: it's not global, it's static :).

File: torch/lib/TH/THStorage.c (hunk 16 above)
Comment (gchanan) [inline torch/lib/TH/THStorage.c]: I don't think they should be consistent -- expand is an error code / success indicator, THSize_isSameSizeAs should return "true" if those are the same size, there's no error code / success in there.

For clarity it's probably just better to define an error code type THError_t.

File: tools/cwrap/plugins/Broadcast.py (hunk 8 above)
Comment (killeent) [inline tools/cwrap/plugins/Broadcast.py]: Yes, I realized this later on.

File: torch/csrc/cuda/expand_utils.cpp (hunk 13 above)
Comment (apaszke) [inline torch/csrc/cuda/expand_utils.cpp]: Can we maybe create a generic dir in utils and generate these specializations via macros? It's going to be much less repetetive

File: torch/csrc/expand_utils.h [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+                    char *to_expand_name, char *tensor_name, bool fallback,
+                    ptrdiff_t to_expand_nElem, ptrdiff_t tensor_nElem, int to_expand_err) {
+  if (fallback && getBackCompatBroadcastWarn()) {
+    bool same_shape = THSize_isSameSizeAs(tensor->size, tensor->nDimension,
+        to_expand->size, to_expand->nDimension);
+    if (!same_shape && to_expand_err == 0 && (tensor_nElem == to_expand_nElem) && fallback) {
+      std::ostringstream warn;
+      warn << tensor_name << " and " << to_expand_name << " do not have the same shape, but are "
+           << "broadcastable, and have the same number of elements.  Changing behavior in a backwards incompatible "
+           << "manner to broadcasting rather than viewing as 1-dimensional.";
```
Comment (apaszke) [inline torch/csrc/expand_utils.h]: Shouldn't we warn even if they are not broadcastable? I thought we're going to deprecate that old behaviour.

File: torch/lib/TH/THSize.h [hunk 18]
```diff
@@ -0,0 +1,13 @@
+#ifndef TH_SIZE_INC
+#define TH_SIZE_INC
+
+#include "THGeneral.h"
+#include <stddef.h>
+
+// THTensor functions that would work on a THSize if we had such a class in C++,
+// i.e. THTensor functions that depend only on the shape of the tensor, not the type.
```
Comment (apaszke) [inline torch/lib/TH/THSize.h]: THTensor functions that depend on the shape, but not type are all `THLongStorage` functions, right? That's what we've used for `inferSize` etc. I think it's a better place for it

File: torch/lib/TH/THStorage.c [hunk 19]
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
...
+      else {
+        THFree(expandedSizes);
+        if (raiseErrors) {
+          THError("The size of tensor a (%ld) must match the size of tensor b (%ld) at "
+                  "non-singleton dimension %ld.", sizeA, sizeB, i);
+        }
+        return -1;
+      }
+    }
+    expandedSizes[ i ] = sizeA;
```
Comment (apaszke) [inline torch/lib/TH/THStorage.c]: wouldn't it be simpler to write it as
```c
if (sizeA == sizeB || sizeA == 1 || sizeB == 1) {
  expandedSizes[i] = THMax(sizeA, sizeB);
} else {
  // Error
}
```

File: torch/lib/TH/THStorage.c [hunk 20]
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
...
+          if (raiseErrors) {
+            THError("The size of tensor %i (%ld) must match the expanded size of tensor (%ld) at "
+                    "non-singleton dimension %ld.", j, size, max_dim_size, i);
+          }
+          return -1;
+        }
+      }
+    }
+
+    expandedSizes[ i ] = max_dim_size;
```
Comment (apaszke) [inline torch/lib/TH/THStorage.c]: as above

File: torch/lib/TH/THStorage.c [hunk 21]
```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
...
+  }
+  THLongStorage_resize(output, ndim);
+  memcpy(THLongStorage_data(output), expandedSizes, sizeof(long)*ndim);
+  THFree(expandedSizes);
+  return 0;
+}
+
+TH_API int THLongStorage_inferExpandGeometry(long *tensorSizes, long *tensorStrides, long tensorDim, THLongStorage *sizes, long **esz, long **est, int raiseErrors) {
```
Comment (apaszke) [inline torch/lib/TH/THStorage.c]: what does this function do? what are esz and est? Expected sizes and expected strides?

File: torch/lib/TH/generic/THTensor.c [hunk 22]
```diff
@@ -285,23 +285,103 @@ void THTensor_(resize5d)(THTensor *self, long size0, long size1, long size2, lon
   THTensor_(resizeNd)(self, 5, size, NULL);
 }
 
-THTensor* THTensor_(newExpand)(THTensor *tensor, THLongStorage *sizes) {
-  THArgCheck(THLongStorage_size(sizes) >= THTensor_(nDimension)(tensor), 1, "the number of sizes provided \
-      must be greater or equal to the number of dimensions in the tensor");
-  THArgCheck(THTensor_(nDimension)(tensor) > 0, 0, "can't expand an empty tensor");
+THTensor* THTensor_(newExpand)(THTensor *tensor, THLongStorage *sizes, int raiseErrors) {
```
Comment (apaszke) [inline torch/lib/TH/generic/THTensor.c]: Why is this new flag necessary? It's not at all obvious why expand functions should have a flag for errors? I think it's better to try to avoid changing the base of tensor API

File: torch/tensor.py [hunk 23]
```diff
@@ -300,7 +302,24 @@ def __matmul__(self, other):
...
+
+            # flatten expanded batches
+            self_expanded = self.expand(*(expand_batch_portion + self_exp_size[-2:])) \
+                .contiguous().view(reduce(mul, expand_batch_portion), *self_exp_size[-2:])
+            other_expanded = other.expand(*(expand_batch_portion + other_exp_size[-2:])) \
+                .contiguous().view(reduce(mul, expand_batch_portion), *other_exp_size[-2:])
+
+            # reshape batches back into result
+            total_expansion = expand_batch_portion + (self_exp_size[-2], other_exp_size[-1])
+            return self_expanded.bmm(other_expanded).view(*(total_expansion))
```
Comment (apaszke) [inline torch/tensor.py]: doesn't `bmm` expand its arguments?

File: torch/utils/backcompat/broadcast/warning/__init__.py [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+
+    def set_enabled(self, value):
+        _set_backcompat_broadcast_warn(value)
+
+    def get_enabled(self):
+        return _get_backcompat_broadcast_warn()
+
+    enabled = property(get_enabled, set_enabled)
+
+sys.modules[__name__] = Warning()
```
Comment (apaszke) [inline torch/utils/backcompat/broadcast/warning/__init__.py]: 1. Don't use old style classes (inherit from `object` explicitly)
2. Wouldn't it be simpler to have a `warning` variable in `torch.utils.backcompat.broadcast` instead of doing this hack?

File: docs/source/notes/broadcasting.rst [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+    >>> y=torch.FloatTensor(3,1,1)
+    # x and y are broadcastable.
+    # 1st trailing dimension: both have size 1
+    # 2nd trailing dimension: y has size 1
+    # 3rd trailing dimension: x has size 1
+    # 4th trailing dimension: y dimension doesn't exist
+    
+    # but:
+    >>> x=torch.FloatTensor(5,2,4,1)
+    >>> y=torch.FloatTensor(3,1,1)
```
Comment (gchanan) [inline docs/source/notes/broadcasting.rst]: sure.

File: test/common_nn.py [hunk 26]
```diff
@@ -103,7 +103,7 @@
     dict(
         module_name='LogSoftmax',
         input_size=(1, 3, 10, 20),
-        reference_fn=lambda i, _: torch.exp(i).div_(torch.exp(i).sum(1).expand_as(i)).log_(),
+        reference_fn=lambda i, _: torch.exp(i).div_(torch.exp(i).sum(1, False).expand_as(i)).log_(),
```
Comment (gchanan) [inline test/common_nn.py]: Nope.

File: torch/_tensor_str.py [hunk 27]
```diff
@@ -67,7 +67,7 @@ def set_printoptions(
 
 def _number_format(tensor, min_sz=-1):
     min_sz = max(min_sz, 2)
-    tensor = torch.DoubleTensor(tensor.nelement()).copy_(tensor).abs_()
+    tensor = torch.DoubleTensor(tensor.size()).copy_(tensor).abs_().resize_(tensor.nelement())
```
Comment (gchanan) [inline torch/_tensor_str.py]: sure.

File: torch/csrc/copy_utils.h (hunk 11 above)
Comment (gchanan) [inline torch/csrc/copy_utils.h]: sure and I'll change the StorageCopy too.

File: torch/csrc/expand_utils-inl.h [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+  return THShortTensor_new();
+}
+
+template <>
+THIntTensor *newForExpand(LIBRARY_STATE_TYPE_NOARGS) {
+  return THIntTensor_new();
+}
+
+template <>
+THLongTensor *newForExpand(LIBRARY_STATE_TYPE_NOARGS) {
```
Comment (gchanan) [inline torch/csrc/expand_utils-inl.h]: We expand on the CPU because the CPU copy methods have THCState if the source is on the GPU.

File: torch/lib/TH/THStorage.c (hunk 21 above)
Comment (gchanan) [inline torch/lib/TH/THStorage.c]: This is just calculateExpandGeometry, I just renamed it to be consistent with other uses of "infer" and simplified the implementation.

To answer your question, I think esz is expandedSizes and est is expandedStrides.  I'll just rename this in this PR, I agree it's confusing.

File: torch/lib/TH/generic/THTensor.c (hunk 22 above)
Comment (gchanan) [inline torch/lib/TH/generic/THTensor.c]: I think I can get rid of it.  I need the raiseErrors flag for other expand functions to support fallback and at one point the Broadcast code called this function, but it doesn't anymore.  So should be safe to remove it here but keep it for expand, expand2, expand3.

File: torch/utils/backcompat/broadcast/warning/__init__.py (hunk 24 above)
Comment (gchanan) [inline torch/utils/backcompat/broadcast/warning/__init__.py]: On 2. I think I did this because properties don't work outside of objects, see https://stackoverflow.com/questions/17007886/property-method-without-class.  At least some quick attempts don't seem to work.

File: torch/lib/TH/generic/THTensor.c (hunk 22 above)
Comment (apaszke) [inline torch/lib/TH/generic/THTensor.c]: Can't you just use a try-catch in THP to handle the cases when it fails?

File: torch/lib/TH/THSize.h (hunk 18 above)
Comment (apaszke) [inline torch/lib/TH/THSize.h]: I think it's better to be consistent. Right now, we use LongStorage for such purposes, and I'd rather keep it like that, with an option to change it as you propose in a later PR.

File: torch/utils/backcompat/broadcast/warning/__init__.py (hunk 24 above)
Comment (apaszke) [inline torch/utils/backcompat/broadcast/warning/__init__.py]: I meant that you could do this instead of hacking `sys.modules`:
```python
# torch/utils/backcompat/broadcast.py
class Warning(object):
    ... # declare the property here
warning = Warning()
```
Also, maybe let's just have `torch.utils.backcompat.broadcast_warning` and `torch.utils.backcompat.keepdim_warning`? You could reuse the same class for this (just give the getter/setter C function to the constructor of both instances).

File: torch/tensor.py (hunk 23 above)
Comment (apaszke) [inline torch/tensor.py]: So does numpy broadcast inputs to matrix multiplication at all?

File: test/test_torch.py [hunk 29]
```diff
@@ -939,6 +944,385 @@ def test_arange(self):
...
+            small2 = None
+            small2Expanded = None
+            if fn in fns_3_args:
+                # create another smaller tensor
+                (dims_small2, _, _) = TestTorch._select_broadcastable_dims(self, dims_full)
+                small2 = torch.randn(*dims_small2).float()
+                small2 = cast(small2)
+                small2Expanded = small2.expand(*dims_full)
+
+            if fn not in fns_no_out_place:
```
Comment (gchanan) [inline test/test_torch.py]: yes, good suggestion.   I was going to see if I could get rid of the exceptions (e.g. just add a max_), but the issue become large enough by itself.

File: test/test_torch.py [hunk 30]
```diff
@@ -939,6 +944,385 @@ def test_arange(self):
...
+
+                # need to clone largeExpanded so we can reuse, since functions are in-place
+                largeExpandedClone = largeExpanded.clone()
+
+                def tensorfn_inplace(t0, t1, t2=None):
+                    t0_fn = getattr(t0, fn + "_")
+                    if fn == "lerp":
+                        return t0_fn(t1, 0.5)
+                    elif fn == "masked_copy":
+                        return t0_fn(t1 < 0.5, t1.expand_as(t0))
```
Comment (gchanan) [inline test/test_torch.py]: No, because it's not really masked_copy, more like masked scatter and the shape of the source has nothing really to do with the shape of the underlying tensor, so we don't broadcast it.

It's just unnecessarily confusing for me to have written the test to use the expanded/non-expanded tensors, though; I'll just use a 1-d arange of the proper length or something.

File: torch/lib/TH/generic/THTensor.c (hunk 22 above)
Comment (gchanan) [inline torch/lib/TH/generic/THTensor.c]: is it guaranteed that THError (what gets called now if raiseErrors = 0) will throw an exception?  The function is settable, right?  So could e.g. exit.

I guess we could move all of this to THPP and add real exception types?

File: torch/tensor.py (hunk 23 above)
Comment (gchanan) [inline torch/tensor.py]: Not for conventional 2-d matrix multiplication, e.g. (7,5) @ (1,3) won't broadcast to (7,5) @ (5,3).

File: torch/lib/TH/THSize.h (hunk 18 above)
Comment (colesbury) [inline torch/lib/TH/THSize.h]: I don't think we should be adding additional uses of THLongStorage to represent sizes. It was a mistake to represent sizes that way that only sort-of made sense in the context of Lua.

We already use `long*` pointers in `resizeNd` and `setStorageNd`.

It's also bad because we take the THLongStorage.size -> THTensor.nDimension, but one is an `int` and the other is a `long`. This requires either `-fpermissive` everywhere or extra casts.

**Comment by gchanan** [comment]
@apaszke I incorporated all your suggestions minus the THLongStorage suggestion (due to the @colesbury comment above) and removing raiseErrors from the expand,expand2,expand3 functions, since that comment thread didn't conclude.

There were also new merge conflicts, so the last commit resolves them via a merge, but let me know if you want me to force push everything via a rebase.

File: torch/lib/TH/generic/THTensor.c (hunk 22 above)
Comment (apaszke) [inline torch/lib/TH/generic/THTensor.c]: Yes, it is guaranteed in PyTorch. We depend on it all over the place (see how `HANDLE_TH_ERRORS` macro is defined in `torch/csrc/Exceptions.h`). I don't think having this flag as part of the interface is a good idea. It's only for this single use case.

**Comment by apaszke** [comment]
I guess we can leave the THSize change, but it'd be better to remove the error flag.

**Comment by gchanan** [comment]
ok, it took some effort to handle the error cases correctly without leaking memory, but I removed the error flag from THTensor.

File: torch/csrc/Module.cpp [hunk 31]
```diff
@@ -485,7 +485,11 @@ PyObject *THPModule_inferSize(PyObject *_unused, PyObject *args)
   THLongStoragePtr sizes_guard(THLongStorage_new());
   THLongStorage *sizes = sizes_guard.get();
 
-  THLongStorage_inferSize2(sizes, size1->data, size1->size, size2->data, size2->size, 1);
+  char error_buffer[1024];
+  int ret = THLongStorage_inferSize2(sizes, size1->data, size1->size, size2->data, size2->size, error_buffer, 1024);
+  if (ret != 0) {
+    THError(error_buffer);
```
Comment (apaszke) [inline torch/csrc/Module.cpp]: It's a python wrapper, you can just use `THPUtils_assert(ret != 0, error_buffer)`. We're not using `THError` anywhere here

**Comment by soumith** [comment]
I fixed the remaining lint and merged this into master!

THIS WAS AWESOME @gchanan

**Comment by soumith** [comment]
i forgot that parts of this PR have TH / THC etc. being touched upon. So I have to do the annoying reverse-merge scheme. So for now I reverted this PR on master (force-pushed, sorry), and I'll merge it in properly tomorrow-ish.

File: torch/tensor.py (hunk 23 above)
Comment (fmassa) [inline torch/tensor.py]: Given that `mm`, `bmm`, `ger` are low-level functions that do not perform broadcasting, what about explicitly stating that in the docs, and add a `torch.matmul` function equivalent to `__matmul__`, with proper documentation? For the moment, the `@` operator is not documented, and only potentially used by py3 users.

Comment (apaszke) [inline torch/tensor.py]: Yeah sounds like a nice thing to have

Comment (gchanan) [inline torch/tensor.py]: Agreed, this is a good idea.  I'm going to open a separate issue for it given this issue is large enough and ready to merge already.

Comment (gchanan) [inline torch/tensor.py]: Filed Issue https://github.com/pytorch/pytorch/issues/1756 for this.

Comment (fmassa) [inline torch/tensor.py]: We probably need to fix the `Variable` case for `__matmul__` to reflect the behavior of the new `__matmul__` as well.

**Comment by soumith** [comment]
this is now properly reverse-merged into master
